User avatars

kiranoot opened the pull request:
Closes #24

Commit:
Add RequestStore gem

Commit:
Add deferred posts system

This allows aggregating posts from many sources and loading them
all at the same time at the end of the request, and only if they
being used.

Other potential approaches are to eager preload all posts referenced
by specific entities, but this breaks for thumbnail dtext, which
will be added later.

Commit:
Enable blacklist code to accept post data as objects

This sets up support of using the blacklist code as a utility to
support blacklisting avatars and dtext thumbnails before inserting
them into the DOM.

Commit:
Add avatar to user class

This also updates the supporting code to enable dumping deferred
post information to the page, which was partially broken in a
previous commit.

Commit:
Thumbnail insertion code

Commit:
Fix preview dimensions, add to applicable pages.

liamwhite commented on app/models/post.rb:
```diff
@@ -1542,6 +1553,31 @@ def as_json(options = {})
+    def minimal_attributes
+      preview_dims = preview_dimensions
+      hash = {
+          'status': status,
```
These keys are symbols, not strings. This is identical to
```rb
hash = {
  status: status,
  flags: status_flags,
  # ...
}
```

If you meant to use strings (I don't think it matters in this case), you would write this:
```rb
hash = {
  'status' => status,
  'flags' => status_flags,
  # ...
}
```

kiranoot replied:
Augh, you're correct. I was thinking of making them into strings, but this is indeed the symbol syntax. I'll just make them proper symbols and be done with it.

liamwhite commented on app/models/post.rb:
```diff
@@ -1542,6 +1553,31 @@ def as_json(options = {})
+      }
+
+      if visible?
+        hash['md5'] = md5
```
Same here (you may have meant `hash[:md5]` instead of `hash['md5']`, etc.)

liamwhite commented on app/javascript/packs/application.js:
```diff
@@ -43,6 +43,7 @@ export { default as Shortcuts } from '../src/javascripts/shortcuts.js';
 export { default as Upload } from '../src/javascripts/uploads.js';
 export { default as Utility } from '../src/javascripts/utility.js';
 export { default as Ugoira } from '../src/javascripts/ugoira.js';
+export { default as Thumbnails } from '../src/javascripts/thumbnails.js';
```
I recognize that this was probably just done for consistency, but the default export of thumbnails.js doesn't seem to be used for anything, so this is equivalent to an import statement loading it into window scope (since it has a window method).

kiranoot replied:
I'll have to double check, but I believe that those exports are to local scope only, and not published into the window object automatically. It was indeed done for consistency, and I have a feeling it's being done this way to avoid linting rules surrounding import for side effects only.

kiranoot replied:
From testing, these are all no-op commands. They both don't push things into the window object, and they don't have any side effects, as they are imported through the importAll command above. I will remove them entirely.

kiranoot replied:
I found out what these are for. It appears that this somehow gets exported into the `Danbooru` namespace which is placed into the window object. So these are still required.

liamwhite commented on app/helpers/application_helper.rb:
```diff
@@ -239,6 +239,14 @@ def data_attributes_for(record, prefix, attributes)
+    post_id = user.avatar_id
+    return "" unless post_id
+    DeferredPosts.add(post_id)
+    tag.div (tag.img class: 'thumb-img placeholder', src: '/images/thumb-preview.png', height: 100, width: 100), class: 'post-thumb placeholder', id: "tp-#{post_id}", 'data-id': post_id
```
I would suggest
```rb
tag.div class: 'post-thumb placeholder', id: "tp-#{post_id}", 'data-id': post_id do
  tag.img class: 'thumb-img placeholder', src: '/images/thumb-preview.png', height: 100, width: 100
end
```
for clarity

kiranoot replied:
👍

liamwhite commented on app/views/static/_deferred_posts.html.erb:
```diff
@@ -0,0 +1,3 @@
+<script>
+  var ___deferred_posts = <%= raw DeferredPosts.dump.to_json %>;
```
Though this is the default behavior for global scripts, I would suggest you explicitly assign this to `window` (`window.__deferredPosts = ...`)

kiranoot replied:
Agreed, since it resolves the ambiguity.

liamwhite commented on app/logical/deferred_posts.rb:
```diff
@@ -0,0 +1,31 @@
+class DeferredPosts
```
This is a bit silly. Here's how I would probably go about this:
```rb
require 'set'

module DeferredPosts
  def deferred_post_ids
    @post_id_set ||= Set.new
  end

  def deferred_posts
    Post.where(id: deferred_post_ids.to_a).find_each.reduce({}) do |post_hash, p|
      post_hash[p.id] = p.minimal_attributes
      post_hash
    end
  end
end
```

Include the module (which for example could be in a [controller concern](https://elegantbrew.tumblr.com/post/70990048275/controller-concerns-in-rails-4)) into the base controller or any subclass thereof:
```rb
  # application_controller.rb
  include DeferredPosts
  helper_method :deferred_post_ids, :deferred_posts
```

Add deferred ids in a controller or view or helper:
```rb
  deferred_post_ids.add(1)
```

...and dump them:
```erb
<script>
  window.__deferredPosts = <%= raw deferred_posts.to_json %>;
</script>
```

No need for request_store or anything of the sort.

kiranoot replied:
I will investigate how this interacts with dtext(the next step, and other user of this functionality) and revise this code based on that experiment. What you suggest looks very reasonable if it works for dtext.

kiranoot replied:
Will fix in the PR for dtext thumbnails.

Commit:
Review changes for user avatars

- Fix confusion between quoted strings and symbols in attributes.
- Explicitly export deferred posts to the window object.
- Remove useless exports of already imported JS classes.
